src/search/merge_and_shrink/transition_system.cc

```diff
@@ -394,42 +466,68 @@ void TransitionSystem::apply_label_reduction(
+            const vector<int> &old_labels = mapping.second;
+            int new_local_label = local_to_global_labels.size();
+            local_to_global_labels.push_back({new_label});
+            local_label_to_cost.push_back(global_labels.get_label_cost(new_label));
+            global_to_local_label[new_label] = new_local_label;
+
+            for (int old_label : old_labels) {
+                int old_local_label = global_to_local_label[old_label];
+                LabelGroup &label_group = local_to_global_labels[old_local_label];
+                label_group.erase(find(label_group.begin(), label_group.end(), old_label));
+                // Reset (for consistency only, old labels are never accessed).
+                global_to_local_label[old_label] = -1;
+            }
+        }
+
+        // Recompute the cost of all affected local labels.
+        for (int local_label : affected_local_labels) {
+            const LabelGroup &label_group = local_to_global_labels[local_label];
+            local_label_to_cost[local_label] = INF;
+            for (int label : label_group) {
+                int cost = global_labels.get_label_cost(label);
+                local_label_to_cost[local_label] = min(local_label_to_cost[local_label], cost);
+            }
+        }
 
         /*
           Go over the transitions of new labels and add them at the correct
           position.
 
           NOTE: it is important that this happens in increasing order of label
-          numbers to ensure that transitions_by_group_id are synchronized with
-          label groups of label_equivalence_relation.
+          numbers to ensure that local_label_to_transitions are synchronized
+          with local_to_global_labels and local_label_to_cost.
         */
         for (size_t i = 0; i < label_mapping.size(); ++i) {
             vector<Transition> &transitions = new_transitions[i];
-            assert(label_equivalence_relation->get_group_id(label_mapping[i].first)
-                   == static_cast<int>(transitions_by_group_id.size()));
-            transitions_by_group_id.push_back(move(transitions));
+            assert(global_to_local_label[label_mapping[i].first]
+                   == static_cast<int>(local_label_to_transitions.size()));
+            local_label_to_transitions.push_back(move(transitions));
         }
 
-        // Go over all affected group IDs and remove their transitions if the
-        // group is empty.
-        for (int group_id : affected_group_ids) {
-            if (label_equivalence_relation->is_empty_group(group_id)) {
-                utils::release_vector_memory(transitions_by_group_id[group_id]);
+        // Finally remove transitions of all local labels that do not represent
+        // any global label anymore.
+        for (int local_label : affected_local_labels) {
+            if (local_to_global_labels[local_label].empty()) {
+                utils::release_vector_memory(local_label_to_transitions[local_label]);
```

Comment:
Could we also do that in the loop that recomputes the cost of all affected local labels? Not sure if we'd want that though...

Reply:
I think we once decided to have every logical step in a separate loop/place. Although you are right, of course.